File: app/src/main/java/com/zaze/demo/debug/AppShortcut.java
package com.zaze.demo.debug;

import android.content.Context;
import android.content.ContextWrapper;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;

import androidx.annotation.NonNull;

import com.zaze.utils.SignaturesUtil;

import org.jetbrains.annotations.NotNull;

/**
 * Description : 应用快捷信息(汇总最新的 Applicaiton 和 packageInfo)
 *
 * @author : ZAZE
 * @version : 2017-12-22 - 17:26
 */
public class AppShortcut {
    private String name;
    private String packageName;
    private int versionCode;
    private String versionName;
    private String sourceDir;
    private int uid;
    private int flags;
    private String signingInfo;
    private boolean isInstalled = false;
    private boolean copyEnable = true;
    private long firstInstallTime = 0;
    private long lastUpdateTime = 0;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getPackageName() {
        return packageName;
    }

    public void setPackageName(String packageName) {
        this.packageName = packageName;
    }

    public int getVersionCode() {
        return versionCode;
    }

    public void setVersionCode(int versionCode) {
        this.versionCode = versionCode;
    }

    public String getVersionName() {
        return versionName;
    }

    public void setVersionName(String versionName) {
        this.versionName = versionName;
    }

    public String getSourceDir() {
        return sourceDir;
    }

    public void setSourceDir(String sourceDir) {
        this.sourceDir = sourceDir;
    }

    public int getUid() {
        return uid;
    }

    public void setUid(int uid) {
        this.uid = uid;
    }

    public int getFlags() {
        return flags;
    }

    public void setFlags(int flags) {
        this.flags = flags;
    }

    public boolean isInstalled() {
        return isInstalled;
    }

    public void setInstalled(boolean installed) {
        isInstalled = installed;
    }

    public String getSigningInfo() {
        return signingInfo;
    }

    public void setSigningInfo(String signingInfo) {
        this.signingInfo = signingInfo;
    }

    public boolean isCopyEnable() {
        return copyEnable;
    }

    public void setCopyEnable(boolean copyEnable) {
        this.copyEnable = copyEnable;
    }

    public long getFirstInstallTime() {
        return firstInstallTime;
    }

    public void setFirstInstallTime(long firstInstallTime) {
        this.firstInstallTime = firstInstallTime;
    }

    public long getLastUpdateTime() {
        return lastUpdateTime;
    }

    public void setLastUpdateTime(long lastUpdateTime) {
        this.lastUpdateTime = lastUpdateTime;
    }

    public static @NonNull
    AppShortcut transform(Context context, final @NonNull PackageInfo packageInfo) {
        AppShortcut appShortcut = new AppShortcut();
        appShortcut.setPackageName(packageInfo.packageName);
        appShortcut.setVersionCode(packageInfo.versionCode);
        appShortcut.setVersionName(packageInfo.versionName);
        appShortcut.setSigningInfo(SignaturesUtil.getSignatures(new ContextWrapper(context) {
            @Override
            public String getPackageName() {
                return packageInfo.packageName;
            }
        }, "MD5"));
        ApplicationInfo applicationInfo = packageInfo.applicationInfo;
        if (applicationInfo != null) {
            appShortcut.setName(context.getPackageManager().getApplicationLabel(applicationInfo).toString());
            appShortcut.setInstalled(true);
            appShortcut.setSourceDir(applicationInfo.sourceDir);
            appShortcut.setFlags(applicationInfo.flags);
            appShortcut.setUid(applicationInfo.uid);
        } else {
            appShortcut.setInstalled(false);
        }
        appShortcut.setFirstInstallTime(packageInfo.firstInstallTime);
        appShortcut.setLastUpdateTime(packageInfo.lastUpdateTime);
        return appShortcut;
    }

    @NotNull
    @Override
    public String toString() {
        return "AppShortcut{" +
                "name='" + name + '\'' +
                ", packageName='" + packageName + '\'' +
                ", versionCode=" + versionCode +
                ", versionName='" + versionName + '\'' +
                ", sourceDir='" + sourceDir + '\'' +
                ", uid=" + uid +
                ", flags=" + flags +
                ", signingInfo='" + signingInfo + '\'' +
                ", isInstalled=" + isInstalled +
                ", copyEnable=" + copyEnable +
                ", firstInstallTime=" + firstInstallTime +
                ", lastUpdateTime=" + lastUpdateTime +
                '}';
    }
}

File: feature/applications/src/main/java/com/zaze/demo/feature/applications/AppShortcut.kt
package com.zaze.demo.feature.applications

import android.app.usage.StorageStatsManager
import android.content.Context
import android.content.ContextWrapper
import android.content.pm.ApplicationInfo
import android.content.pm.PackageInfo
import android.content.res.Resources
import android.graphics.Bitmap
import android.os.Build
import android.os.storage.StorageManager
import androidx.annotation.RequiresApi
import com.google.gson.annotations.Expose
import com.zaze.utils.AppUtil
import com.zaze.utils.FileUtil
import com.zaze.utils.SignaturesUtil
import com.zaze.utils.ext.getLabel
import com.zaze.utils.log.ZLog
import com.zaze.utils.log.ZTag
import java.lang.Exception
import java.util.*


/**
 * Description : 应用快捷信息(汇总最新的 Applicaiton 和 packageInfo)
 *
 * @author : ZAZE
 * @version : 2017-12-22 - 17:26
 */
data class AppShortcut(

    val appName: String? = null,
    val packageName: String = "",
    val versionCode: Long = 0,
    val versionName: String? = null,
    val uid: Int = 0,
    val flags: Int = 0,
    val isInstalled: Boolean = false,
    val firstInstallTime: Long = 0,
    val lastUpdateTime: Long = 0,
    val installerPackageName: String? = null,
    @Transient
    @Expose(serialize = false, deserialize = false)
    val packageInfo: PackageInfo? = null,
    @Transient
    @Expose(serialize = false, deserialize = false)
    val applicationInfo: ApplicationInfo? = null
) {
    // --------------------------------------------------
    var isLocalFile: Boolean = false

    val enable: Boolean get() = applicationInfo?.enabled ?: false

    val sourceDir: String? = applicationInfo?.sourceDir

    val apkSize: Long
        get() = FileUtil.getFileSize(applicationInfo?.sourceDir)

    @Transient
    @Expose(serialize = false, deserialize = false)
    private var appIcon: Bitmap? = null

    fun updateAppIcon(bitmap: Bitmap) {
        appIcon = bitmap
    }

    fun getAppIcon(context: Context? = null): Bitmap? {
        if (appIcon == null && context != null) {
            // 默认不赋值，每次都从ApplicationManager缓存中获取
            return ApplicationManager.getAppIconHasDefault(context, packageName)
        }
        return appIcon
    }

    companion object {
        fun empty(packageName: String): AppShortcut {
            return AppShortcut(packageName = packageName)
        }

        @JvmStatic
        fun create(context: Context, packageInfo: PackageInfo): AppShortcut {
            val applicationInfo = packageInfo.applicationInfo
            val isInstalled = applicationInfo != null
            val appShortcut = AppShortcut(
                packageName = packageInfo.packageName,
                versionCode = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                    packageInfo.longVersionCode
                } else {
                    packageInfo.versionCode.toLong()
                },
                versionName = packageInfo.versionName,
                appName = applicationInfo?.getLabel(context),
                isInstalled = isInstalled,
                flags = applicationInfo?.flags ?: 0,
                uid = applicationInfo?.uid ?: 0,
                firstInstallTime = packageInfo.firstInstallTime,
                lastUpdateTime = packageInfo.lastUpdateTime,
                installerPackageName = if (isInstalled) AppUtil.getInstallerPackageName(
                    context,
                    packageInfo.packageName
                ) else null,
                packageInfo = packageInfo,
                applicationInfo = applicationInfo
            )
//            ZLog.i(ZTag.TAG, "appShortcut: $appShortcut")
            return appShortcut
        }
    }

    fun isSystemApp(): Boolean {
        return flags and ApplicationInfo.FLAG_SYSTEM != 0
    }

    fun getSignatures(context: Context): String? {
        return if (!isLocalFile) {
            SignaturesUtil.getSignatures(
                AppUtil.getSignatures(context, this.packageName),
                "MD5"
            )
        } else {
            SignaturesUtil.getSignatures(
                AppUtil.getApkFileSignatures(context, this.sourceDir),
                "MD5"
            )
        }
    }

    @RequiresApi(Build.VERSION_CODES.O)
    fun queryStorageStats(context: Context) {
        val storageManager = context.getSystemService(Context.STORAGE_SERVICE) as StorageManager
        val storageStatsManager =
            context.getSystemService(Context.STORAGE_STATS_SERVICE) as StorageStatsManager
        storageManager.storageVolumes.forEach { volume ->
            val uuid = volume.uuid?.let {
                UUID.fromString(it)
            } ?: StorageManager.UUID_DEFAULT
            val storageStats = storageStatsManager.queryStatsForPackage(
                uuid,
                packageName,
                android.os.Process.myUserHandle()
            )
            ZLog.i(
                ZTag.TAG,
                "$appName ${volume.getDescription(context)} appBytes: ${storageStats.appBytes}; cacheBytes: ${storageStats.cacheBytes}; dataBytes: ${storageStats.dataBytes}"
            )
        }
    }

    fun getResource(context: Context): Resources? {
        return applicationInfo?.let {
            try {
                context.packageManager.getResourcesForApplication(it)
            } catch (e: Exception) {
                null
            }
        }
    }
}